README.md
<img src="art/fake-logo.png" alt="syft.js logo" width="200" />

![Build](https://img.shields.io/github/workflow/status/OpenMined/syft.js/Run%20tests%20and%20coverage)
![codecov](https://img.shields.io/codecov/c/github/OpenMined/syft.js)
![npm](https://img.shields.io/npm/v/@openmined/syft.js)
![GitHub](https://img.shields.io/github/license/OpenMined/syft.js.svg)
![OpenCollective](https://img.shields.io/opencollective/all/openmined)

<!-- ALL-CONTRIBUTORS-BADGE:START - Do not remove or modify this section -->

[![All Contributors](https://img.shields.io/badge/all_contributors-8-orange.svg?style=flat-square)](#contributors-)

<!-- ALL-CONTRIBUTORS-BADGE:END -->

# Syft.js

Syft.js is the “web” part of the [OpenMined](https://openmined.org)'s open-source ecosystem for [federated learning](https://ai.googleblog.com/2017/04/federated-learning-collaborative.html),
which currently spans across web, [iOS](https://github.com/OpenMined/SwiftSyft), [Android](https://github.com/OpenMined/KotlinSyft), and [servers/IoT](https://github.com/OpenMined/PySyft).

Syft.js has following core features:

- :hammer_and_wrench: Integration with PyGrid federated learning API.
- :gear: **Training and inference** of any PySyft model written in PyTorch or TensorFlow.
- :bust_in_silhouette: Allows all data to stay on the user's device.
- :lock: Support for **secure multi-party computation** and **secure aggregation** protocols using **peer-to-peer WebRTC** connections (_in progress_).

The library is built on top of [TensorFlow.js](https://js.tensorflow.org/).

There are a variety of additional privacy-preserving protections that may be applied, including [differential privacy](https://towardsdatascience.com/understanding-differential-privacy-85ce191e198a), [muliti-party computation](https://www.inpher.io/technology/what-is-secure-multiparty-computation), and [secure aggregation](https://research.google/pubs/pub45808/).

If you want to know how scalable federated systems are built, [Towards Federated Learning at Scale](https://arxiv.org/pdf/1902.01046.pdf) is a fantastic introduction!

## Installation

Note that syft.js needs Tensorflow.js library as peer dependency.

If you're using a package manager like NPM:

```text
npm install --save @openmined/syft.js @tensorflow/tfjs-core
```

Or if Yarn is your cup of tea:

```text
yarn add @openmined/syft.js @tensorflow/tfjs-core
```

If you're not using a package manager, you will be able to include Syft.js within a `<script>` tag.
In this case library classes will be available under `syft` global object.

```html
<script src="https://cdn.jsdelivr.net/npm/@tensorflow/tfjs@1.2.5/dist/tf.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/@openmined/syft.js@latest/dist/index.js"></script>

<script type="text/javascript">
  // Create syft worker
  const worker = syft.Syft({...});
  ...
</script>
```

## Quick Start

As a developer, there are few steps to building your own secure federated learning system upon the OpenMined infrastructure:

1. :robot: [Develop ML model and training procedure](https://github.com/OpenMined/PySyft/blob/master/examples/tutorials/model-centric-fl/Part%2001%20-%20Create%20Plan.ipynb) (aka `Plan` in PySyft terminology) using [PySyft](https://github.com/OpenMined/PySyft).
1. :earth_americas: Host model and Plans on [PyGrid](https://github.com/OpenMined/PyGrid), which will deal with all the federated learning components of your pipeline.
1. :tada: Execute the training on the variety of end-user devices using the client library (syft.js, [SwiftSyft](https://github.com/OpenMined/SwiftSyft), [KotlinSyft](https://github.com/OpenMined/KotlinSyft), [PySyft](https://github.com/OpenMined/PySyft)).
1. :lock: Securely aggregate trained user models in PyGrid.

**:notebook: The entire workflow and process is described in greater detail in the [Web & Mobile Federated Learning project roadmap](https://github.com/OpenMined/Roadmap/blob/master/web_and_mobile_team/projects/federated_learning.md).**

Syft.js provides minimalistic API to communicate with federated learning PyGrid endpoints
and execute PySyft's Plans in a browser.
The federated learning cycle implemented with syft.js would contain following steps:

- Register into training cycle on PyGrid.
- Download required model and Plans from PyGrid.
- Execute the Plan with given model parameters and local user's data.
- Submit difference between original and trained model parameters for aggregation.

This whole cycle can be expressed in the following code:

```javascript
import * as tf from '@tensorflow/tfjs-core';
import { Syft } from '@openmined/syft.js';

const gridUrl = 'ws://pygrid.myserver.com:5000';
const modelName = 'my-model';
const modelVersion = '1.0.0';

// if the model is protected with authentication token (optional)
const authToken = '...';

const worker = new Syft({ gridUrl, authToken, verbose: true });
const job = await worker.newJob({ modelName, modelVersion });
job.start();

job.on('accepted', async ({ model, clientConfig }) => {
  const batchSize = clientConfig.batch_size;
  const lr = clientConfig.lr;

  // Load data.
  const batches = LOAD_DATA(batchSize);

  // Load model parameters.
  let modelParams = model.params.map((p) => p.clone());

  // Main training loop.
  for (let [data, labels] of batches) {
    // NOTE: this is just one possible example.
    // Plan name (e.g. 'training_plan'), its input arguments and outputs depends on FL configuration and actual Plan implementation.
    let updatedModelParams = await job.plans['training_plan'].execute(
      job.worker,
      data,
      labels,
      batchSize,
      lr,
      ...modelParams
    );

    // Use updated model params in the next iteration.
    for (let i = 0; i < modelParams.length; i++) {
      modelParams[i].dispose();
      modelParams[i] = updatedModelParams[i];
    }
  }

  // Calculate & send model diff.
  const modelDiff = await model.createSerializedDiff(modelParams);
  await job.report(modelDiff);
});

job.on('rejected', ({ timeout }) => {
  // Handle the job rejection, e.g. re-try after timeout.
});

job.on('error', (err) => {
  // Handle errors.
});
```

Note that syft.js doesn't handle user's data collection, data storage and loading.

### API Documentation

See [API Documentation](API-REFERENCE.md) for complete reference.

## Running the Demo App

The “Hello World” syft.js demo is MNIST training example located in `examples/mnist` folder.
It demonstrates how a simple neural net model created in PySyft can be trained in a browser
and the result of training averaged from multiple federated learning participants.

<img src="art/mnist-demo-ani.gif" alt="syft.js MNIST demo animation" />

Running the demo is multi-stage and multi-component process
(as the federated learning itself).

Below are example instructions that assume you
want to put everything under `~/fl-demo` folder.

### Installation

It is recommended that you install python packages in separate virtualenv or conda environment, e.g.:

```bash
virtualenv -p python3 syft
source syft/bin/activate
```

or

```bash
conda create -n syft python=3.7
conda activate syft
```

Now, you will need to install following packages:

- PySyft. Follow
  [PySyft installation guide](https://github.com/OpenMined/PySyft/blob/master/INSTALLATION.md)
  to install the latest `master` branch of PySyft.

- PyGrid.
  Follow [PyGrid documentation](https://github.com/OpenMined/PyGrid/#start-grid-platform-locally)
  to install the latest `dev` branch of PyGrid.
  NOTE: If you are running PyGrid with docker-compose, you'll need to [build docker images](https://github.com/OpenMined/PyGrid/#13---build-your-own-images-optional)
  from the latest `dev` branch.

- Syft.js with MNIST demo. Check out the latest `master` branch of syft.js with MNIST demo app included:
  ```bash
  cd ~/fl-demo
  git clone https://github.com/OpenMined/syft.js
  cd syft.js
  npm install
  cd examples/mnist
  npm install
  ```

#### Seeding the Model & Plan

Syft.js connects to PyGrid to pick up the model and training Plan.
For the demo to work, we need to populate that data into PyGrid.

##### Run PyGrid

There're two possible ways to start PyGrid:

- Run `./dev_server.sh` script in the PyGrid root folder.
- Run `docker-compose up --build` in the PyGrid root folder.

Here we assume you don't need to change default PyGrid configuration and it listens
on the `localhost:5000`. If you need to use different host/port,
PyGrid URL will need to be adjusted accordingly in further steps.

##### Create Model & Plan

After PyGrid is running, the next step is to create the model and training plan and host them in PyGrid.
PySyft tutorials include [MNIST example jupyter notebooks](https://github.com/OpenMined/PySyft/blob/master/examples/tutorials/model-centric-fl) that guide you through this process.

Fire up jupyter notebook in PySyft root folder:

```bash
cd ~/fl-demo/PySyft
jupyter notebook --notebook-dir=$(pwd)
```

In the console, you should see URL you should open, or the browser will open automatically. After this, navigate to `examples/tutorials/model-centric-fl` and [run the first notebook](https://github.com/OpenMined/PySyft/blob/master/examples/tutorials/model-centric-fl/Part%2001%20-%20Create%20Plan.ipynb). At this point, you can pull down the model and training plan with syft.js. However, if you'd like to see how to execute the plan using the PySyft FL worker, [try running the second notebook](https://github.com/OpenMined/PySyft/blob/master/examples/tutorials/model-centric-fl/Part%2002%20-%20Execute%20Plan.ipynb).

##### PyGrid Clean-up

In case you need to reset PyGrid database to blank state, stop the process with `Ctrl+C` and remove `databaseGateway.db` file in PyGrid.
Or, if you used docker-compose, stop and re-start it using `docker-compose up --force-recreate` command.

#### Starting the Demo

Finally, we got to the browser part of the demo:

```bash
cd ~/fl-demo/syft.js/examples/mnist
npm start
```

This should start development server and open `localhost:8080` in the browser.
Assuming PyGrid URL, MNIST model name and version were not modified in previous steps, just
press “Start FL Worker”.

You should see following in dev console:

- Syft.js registers into training cycle on PyGrid and gets configuration, Plan, and the model.
- App loads MNIST dataset and executes the training plan with each data batch.
  Charts are updated during this process, and you should see the training loss going down and the accuracy going up.
- After the training is complete, model diff is submitted to PyGrid.

If “Keep making cycle requests” is checked, the whole cycle process is repeated until PyGrid tells worker that model training is complete.

## Compatibility

### PySyft

Syft.js has been tested with PySyft 0.2.7

### PyGrid

Syft.js has been tested with the latest version of PyGrid on `master`.

### Tensorflow.js

Syft.js was tested with Tensorflow.js v1.2.5.

### Browser Support

Syft.js was tested with Chrome and Firefox browsers.

## Support

For support in using this library, please join the **#lib_syft_js** Slack channel. If you’d like to follow along with any code changes to the library, please join the **#code_syftjs** Slack channel. [Click here to join our Slack community!](https://slack.openmined.org)

## Contributing

Please check [open issues](https://github.com/OpenMined/syft.js/issues) as a starting point.

Bug reports and feature suggestions are welcomed as well.


CHANGELOG.md
### Changelog

All notable changes to this project will be documented in this file. Dates are displayed in UTC.

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

#### [v0.1.4](https://github.com/OpenMined/syft.js/compare/v0.1.3...v0.1.4)

> 11 August 2020

- Test - src/sockets.js [`#209`](https://github.com/OpenMined/syft.js/pull/209)
- Documentation - src/types/placeholder.js [`#208`](https://github.com/OpenMined/syft.js/pull/208)
- Documentation: src/types/message.js [`#207`](https://github.com/OpenMined/syft.js/pull/207)
- Avoid Buffer and crypto dependencies [`#211`](https://github.com/OpenMined/syft.js/pull/211)
- Documentation: src/sockets [`#201`](https://github.com/OpenMined/syft.js/pull/201)
- Added documentation and testing for object-registry.js [`#206`](https://github.com/OpenMined/syft.js/pull/206)
- Documentation: src/syft-model.js [`#199`](https://github.com/OpenMined/syft.js/pull/199)
- Documentation: src/job.js [`#197`](https://github.com/OpenMined/syft.js/pull/197)
- doc: update changelog [`5d57e32`](https://github.com/OpenMined/syft.js/commit/5d57e321fea060ad54fe849d3b58647ddaba4320)
- Align dev with published 0.1.3 version [`ef2441e`](https://github.com/OpenMined/syft.js/commit/ef2441e1e8efb5d38448d73d49a93c17fe98355d)

#### [v0.1.3](https://github.com/OpenMined/syft.js/compare/v0.1.2...v0.1.3)

> 29 July 2020

- Documentation: Edit src/syft.js [`#195`](https://github.com/OpenMined/syft.js/pull/195)
- Changing the routes again [`ce45dee`](https://github.com/OpenMined/syft.js/commit/ce45dee8100050a369e2b2479e7d1703bd488675)
- Changing static to model-centric [`06e1bf9`](https://github.com/OpenMined/syft.js/commit/06e1bf943ec9c50ee1aa32e170f7ec9e16eaf0cc)

#### [v0.1.2](https://github.com/OpenMined/syft.js/compare/v0.1.1...v0.1.2)

> 22 July 2020

- Fix speed test edge case [`#192`](https://github.com/OpenMined/syft.js/pull/192)
- Make pygrid speed test optional [`#193`](https://github.com/OpenMined/syft.js/pull/193)
- Removing confusing changelog [`fb70437`](https://github.com/OpenMined/syft.js/commit/fb704373130c1110ae9a268b93bd0fc2d44e0e84)

#### [v0.1.1](https://github.com/OpenMined/syft.js/compare/v0.1.0...v0.1.1)

> 22 July 2020

- docs: add IamRavikantSingh as a contributor [`#187`](https://github.com/OpenMined/syft.js/pull/187)
- Fix typo [`#184`](https://github.com/OpenMined/syft.js/pull/184)
- bandit hookup wip [`7ffa98f`](https://github.com/OpenMined/syft.js/commit/7ffa98f8ac6276177d1d349b5d42dc70c7d2042a)
- Initial cleanup of bandit code [`e4df511`](https://github.com/OpenMined/syft.js/commit/e4df5115b999e338044abb475523b3c960e0321a)
- Updating minor and patch deps [`69ec08c`](https://github.com/OpenMined/syft.js/commit/69ec08ccac6e492e1179eabcb022acd63fe53ea5)

#### [v0.1.0](https://github.com/OpenMined/syft.js/compare/v0.0.1-1...v0.1.0)

> 10 July 2020

- docs: add tsingh2k15 as a contributor [`#181`](https://github.com/OpenMined/syft.js/pull/181)
- docs: add Benardi as a contributor [`#183`](https://github.com/OpenMined/syft.js/pull/183)
- docs: add pedroespindula as a contributor [`#182`](https://github.com/OpenMined/syft.js/pull/182)
- docs: add vkkhare as a contributor [`#180`](https://github.com/OpenMined/syft.js/pull/180)
- docs: add IamRavikantSingh as a contributor [`#179`](https://github.com/OpenMined/syft.js/pull/179)
- Add tests with bandit plans [`#176`](https://github.com/OpenMined/syft.js/pull/176)
- docs: add Nolski as a contributor [`#178`](https://github.com/OpenMined/syft.js/pull/178)
- docs: add vvmnnnkv as a contributor [`#177`](https://github.com/OpenMined/syft.js/pull/177)
- Update readme to reference pysyft, pygrid installation docs [`#175`](https://github.com/OpenMined/syft.js/pull/175)
- Migrate to TFJS-traslated plans [`#174`](https://github.com/OpenMined/syft.js/pull/174)
- Fix README after allcontributors bot changes [`#164`](https://github.com/OpenMined/syft.js/pull/164)
- Add PyGrid integration tests [`#170`](https://github.com/OpenMined/syft.js/pull/170)
- Add Support section to README.md [`#168`](https://github.com/OpenMined/syft.js/pull/168)
- MNIST example: use syft.js remote dependency instead of local [`#165`](https://github.com/OpenMined/syft.js/pull/165)
- docs: add cereallarceny as a contributor [`#163`](https://github.com/OpenMined/syft.js/pull/163)
- Replace bundling crypto with specific randomfill package. Fixes #161 [`#162`](https://github.com/OpenMined/syft.js/pull/162)
- Update README [`#157`](https://github.com/OpenMined/syft.js/pull/157)
- Fix error when model params are nn.Parameter [`#156`](https://github.com/OpenMined/syft.js/pull/156)
- Generate API docs [`#153`](https://github.com/OpenMined/syft.js/pull/153)
- Rename grid.js example to mnist, and do small cleanup [`#152`](https://github.com/OpenMined/syft.js/pull/152)
- Make up-to-date with protobuf and Plan changes in PySyft [`#151`](https://github.com/OpenMined/syft.js/pull/151)
- Add Speed Test [`#117`](https://github.com/OpenMined/syft.js/pull/117)
- Remove .travis.yml [`#118`](https://github.com/OpenMined/syft.js/pull/118)
- Update protobuf types and Plan, temporarily rollback Threepio usage [`#114`](https://github.com/OpenMined/syft.js/pull/114)
- Threepio integration [`#112`](https://github.com/OpenMined/syft.js/pull/112)
- Wire with PyGrid report [`#110`](https://github.com/OpenMined/syft.js/pull/110)
- Secure web-socket connection for GridAPIClient [`#111`](https://github.com/OpenMined/syft.js/pull/111)
- Add singleton support for Logger class [`#106`](https://github.com/OpenMined/syft.js/pull/106)
- Forcing all websocket connections to secure if not debug mode. [`#107`](https://github.com/OpenMined/syft.js/pull/107)

src/Model.ts
// import * as Async from 'promasync'
import * as controller from './controller'
import { Optimizer } from './Optimizer'
import { assertType } from './asserts'
import {
  Tensor,
  IntTensor,
  FloatTensor
} from './Tensor'

import {
  AsyncInstance,
  IAsyncConstructor
} from './AsyncClass'

export class Model extends AsyncInstance {
  type = 'model'
  layerType = '(unknown)'
  params: boolean
  outputShape?: number|string = '(dynamic)'

  protected static assertLayerType(
    a: string,
    b: Function
  ) {
    if (a.toLowerCase() !== b.name.toLowerCase()) {
      throw new TypeError(`Connat Convert '${a}' to '${b.name}'`)
    }
  }

  protected static newModel(
    $: any,
    id: string,
    type: string
  ): Model {
    AsyncInstance.assertCallable($)

    switch(type) {
      case 'policy':
        return new Policy(AsyncInstance, id)
      case 'sequential':
        return new Sequential(AsyncInstance, id)
      case 'linear':
        return new Linear(AsyncInstance, id)
      case 'relu':
        return new ReLU(AsyncInstance, id)
      case 'dropout':
        return new Dropout(AsyncInstance, id)
      case 'sigmoid':
        return new Sigmoid(AsyncInstance, id)
      case 'softmax':
        return new Softmax(AsyncInstance, id)
      case 'logsoftmax':
        return new LogSoftmax(AsyncInstance, id)
      case 'log':
        return new Log(AsyncInstance, id)
      case 'tanh':
        return new Tanh(AsyncInstance, id)
      case 'mseloss':
        return new MSELoss(AsyncInstance, id)
      case 'nllloss':
        return new NLLLoss(AsyncInstance, id)
      case 'crossentropyloss':
        return new CrossEntropyLoss(AsyncInstance, id)
      case 'categorical_crossentropy':
        return new Categorical_CrossEntropy(AsyncInstance, id)
      default:
        throw new Error(`Unkown Model Type: ${type}`)
    }
  }

  static async getModelType(
    id: string
  ): Promise<string> {
    return assertType(
      await controller.sendJSON({
        functionCall: 'model_type',
        objectType: 'model',
        objectIndex: id,
        tensorIndexParams: []
      }, 'string'),
      'string'
    ) as string
  }

  static async getModel(
    id: string
  ): Promise<Model> {
    let type = await Model.getModelType(id)

    return Model.newModel(AsyncInstance, id, type)
  }

  static async createModel(
    layerConstructor: Function,
    ...params: any[] // TODO: what type are thses
  ) {
    let layerType = layerConstructor.name.toLowerCase()

    return assertType(
      await controller.sendJSON({
        functionCall: 'create',
        objectType: 'model',
        tensorIndexParams: [layerType, ...params]
      }, 'string'),
      'string'
    ) as string
  }

  async feed(...args: any[]) {
    let self = this
    self.ready()

    return self.forward(...args)
  }

  async parameters(): Promise<Tensor[]> {
    let self = this
    self.ready()

    return assertType(
      await controller.sendJSON(self.cmd({
        functionCall: 'params'
      }), 'FloatTensor_list'),
      Array
    )
  }

  async num_parameters() {
    let self = this
    self.ready()

    return assertType(
      await controller.sendJSON(self.cmd({
        functionCall: 'param_count'
      }), 'int'),
      'number'
    )
  }

  async models(): Promise<Model[]> {
    let self = this
    self.ready()

    return assertType(
      await controller.sendJSON(self.cmd({
        functionCall: 'models'
      }), 'Model_list'),
      Array
    )
  }

  async set_id(
    new_id: string
  ) {
    let self = this
    self.ready()

    assertType(
      await controller.sendJSON(self.cmd({
        functionCall: 'set_id',
        tensorIndexParams: [new_id]
      }), 'string'),
      'string'
    )

    self.id = new_id
    return self
  }

  async fit(
    input: Tensor,
    target: Tensor,
    criterion: Model,
    optim: Optimizer,
    batch_size: number,
    iters = 15,
    log_interval = 200,
    metrics: string[] = [],
    verbose = true
  ) {
    let self = this
    self.ready()

    console.log('fit')

    let num_batches = assertType(
      await controller.sendJSON(self.cmd({
        functionCall: 'prepare_to_fit',
        tensorIndexParams: [input.id, target.id, criterion.id, optim.id, batch_size]
      }), 'int'),
      'number'
    )

    let loss = 100000
    for (let iter = 0; iter < iters; iter++) {
      for (let log_i = 0; log_i < num_batches; log_i += log_interval) {
        let prev_loss = loss

        let _loss = assertType(
          await controller.sendJSON(self.cmd({
            functionCall: 'fit',
            tensorIndexParams: [log_i, Math.min(log_i + log_interval, num_batches), 1]
          }), 'float'),
          'number'
        ) as number

        if (log_i % 10 === 0) {
          console.log(`iter ${iter}/${iters} - ${log_i}/${num_batches} -- ${_loss}`)
        }

        if(_loss) {
          loss = _loss
        } else {
          console.log(_loss)
        }
        if (Number.isNaN(loss) || Number.isNaN(prev_loss)) {
          break
        }
      }
      if (Number.isNaN(loss)) {
        break
      }
    }
    return loss
  }
  //
  // async summary(
  //   verbose = true,
  //   return_instead_of_print = false
  // ): Promise<string|undefined> {
  //   let self = this
  //   self.ready()
  //
  //   // let layerType = await self.getLayerType() + '_' + self.id + ' (' + str(type()).split('\'')[1].split('.')[-1] + ')'
  //   let layerType = `${await self.getLayerType()}_${self.id} (${self.type})`
  //   let outputShape = ''
  //   if (typeof self.outputShape === 'number') {
  //     outputShape = String(self.outputShape)
  //   } else {
  //     outputShape = String(self.outputShape)
  //   }
  //
  //   let n_param = String(await self.num_parameters())
  //   let output = layerType + ' '.repeat(29 - layerType.length) + outputShape + ' '.repeat(26 - outputShape.length) + n_param + '\n'
  //   if (verbose) {
  //     let single = '_________________________________________________________________\n'
  //     let header = 'Layer (type)                 Output Shape              Param #   \n'
  //     let double = '=================================================================\n'
  //     // TODO: let total_params = 'Total params: ' + '{:,}'.format(self.num_parameters()) + '\n'
  //     // TODO: let trainable_params = 'Trainable params: ' + '{:,}'.format(self.num_parameters()) + '\n'
  //     let non_trainable_params = 'Non-trainable params: 0' + '\n'
  //     // TODO: output = single + header + double + output + double + total_params + trainable_params + non_trainable_params + single
  //   }
  //
  //   if (return_instead_of_print) {
  //     return output
  //   }
  //   console.log(output)
  //   return
  // }

  async length() {
    let self = this
    self.ready()

    return (await self.models()).length
  }


  async activation() {
    let self = this
    self.ready()

    return assertType(
      await controller.sendJSON(self.cmd({
        functionCall: 'activation'
      }), 'FloatTensor', /*delete_after_use=false*/),
      FloatTensor
    )
  }

  async getLayerType() {
    let self = this
    self.ready()

    return assertType(
      await controller.sendJSON(self.cmd({
        functionCall: 'model_type'
      }), 'string'),
      'string'
    )
  }

  cmd(
    options: {
      [key: string]: any
      functionCall: string
      tensorIndexParams?: any[],
    }
  ): SocketCMD {
    let self = this

    return {
      objectType: self.type,
      objectIndex: self.id || '-1',
      tensorIndexParams: [],
      ...options
    }
  }

  async forward(
    ...input: Tensor[]
  ) {
    let self = this
    self.ready()

    return assertType(
      await controller.sendJSON(self.cmd({
        functionCall: 'forward',
        tensorIndexParams: input.map(t => t.id)
      }), 'FloatTensor' /*, false*/),
      FloatTensor
    )
  }

  // async __repr__(
  //   verbose = true
  // ) {
  //   let self = this
  //   self.ready()
  //
  //   if (verbose) {
  //     let output = ''
  //     output += self.__repr__(false) + '\n'
  //     for (let p of await self.parameters()) {
  //       output += '\t W:' + p.__repr__(false)
  //     }
  //     let activation = await self.activation()
  //     if (activation) {
  //       output += '\t A:' + activation + '\n'
  //     }
  //     return output
  //   } else {
  //     return `<syft.nn.${self.layerType} at ${self.id}>`
  //   }
  // }
}

export class Policy extends Model {
  static $: IAsyncConstructor = Policy
  layerType = 'policy'
  stateType: string
  optimizer?: Optimizer
  model?: Model

  static async get(
    id: string
  ) {
    let type = await Model.getModelType(id)

    Model.assertLayerType(type, this)

    return new this(AsyncInstance, id)
  }

  static async create(
    model: Model,
    optimizer: Optimizer,
    stateType = 'discrete'
  ) {
    let id = await Model.createModel(this, model.id, optimizer.id)

    let policy = new this(AsyncInstance, id)

    policy.stateType = stateType

    return policy
  }

  async sample(
    ...input: Tensor[]
  ) {
    let self = this
    self.ready()

    return assertType(
      await controller.sendJSON(self.cmd({
        functionCall: 'sample',
        tensorIndexParams: input.map(t => t.id)
      }), 'IntTensor'),
      IntTensor
    )
  }

  async parameters() {
      let self = this
      self.ready()

    if (self.model) {
      return self.model.parameters()
    }

    return []
  }

  async feed(
    ...args: any[]  // TODO: what type is this
  ) {
    let self = this
    self.ready()

    if (self.stateType === 'discrete') {
      return self.sample(...args)
    } else if (self.stateType === 'continuous') {
      return self.forward(...args)
    }

    throw new Error(`Unknown State Type: ${self.stateType}`)
  }

  // async history() {
  //     let self = this
  //     self.ready()
  //
  //   let raw_history = await controller.sendJSON(self.cmd({
  //     functionCall: 'get_history'
  //   }), 'string')
  //   // TODO: let history_idx = list(map(lambda x:list(map(lambda y:int(y),x.split(','))),raw_history[2:-1].split('],[')))
  //   let losses = []
  //   let rewards = []
  //
  //   for (let {loss, reward} of history_idx) {
  //     if (loss !== -1) {
  //       losses.push(await controller.get_tensor(loss))
  //     } else {
  //       losses.push(void 0)
  //     }
  //     if (reward !== -1) {
  //       rewards.push(await controller.get_tensor(reward))
  //     } else {
  //       rewards.push(void 0)
  //     }
  //   }
  //
  //   return [losses, rewards]
  // }
}

export class Sequential extends Model {
  static $: IAsyncConstructor = Sequential
  layerType = 'sequential'

  static async get(
    id: string
  ) {
    let type = await Model.getModelType(id)

    Model.assertLayerType(type, this)

    return new this(AsyncInstance, id)
  }

  static async create(
    layers?: Model[]
  ) {
    let id = await Model.createModel(this)

    let model = new this(AsyncInstance, id)

    if (Array.isArray(layers)) {
      for (let layer of layers) {
        await model.add(layer)
      }
    }

    return model
  }

  async add(
    model: Model
  ) {
    let self = this
    self.ready()

    await controller.sendJSON(self.cmd({
      functionCall: 'add',
      tensorIndexParams: [model.id]
    }), /*delete_after_use=false*/)
  }

  // async summary() {
  //   let self = this
  //   self.ready()
  //
  //   let single = '_________________________________________________________________\n'
  //   let header = 'Layer (type)                 Output Shape              Param #   \n'
  //   let double = '=================================================================\n'
  //   // TODO: let total_params = 'Total params: ' + '{:,}'.format(self.num_parameters()) + '\n'
  //   // TODO: let trainable_params = 'Trainable params: ' + '{:,}'.format(self.num_parameters()) + '\n'
  //   let non_trainable_params = 'Non-trainable params: 0' + '\n'
  //
  //   let output = single + header + double
  //   Async.each
  //
  //   let mods = await Async.map(await self.models() as Model[], async (m) => {
  //     return m.summary(false, true)
  //   })
  //   output += mods.join(single)
  //   output += double
  //   // TODO: output += total_params + trainable_params + non_trainable_params + single
  //   console.log(output)
  //   return output
  // }
}

export class Linear extends Model {
  static $: IAsyncConstructor = Linear
  layerType = 'linear'

  static async get(
    id: string
  ) {
    let type = await Model.getModelType(id)

    Model.assertLayerType(type, this)

    return new this(AsyncInstance, id)
  }

  static async create(
    input_dim = 0,
    output_dim = 0,
    initializer = 'Xavier'
  ) {
    let id = await Model.createModel(this, String(input_dim), String(output_dim), initializer)

    return new this(AsyncInstance, id)
  }

  async finish(
    id: string
  ) {
    let self = this

    self.id = id

    // let params = await self.parameters()

    // TODO: self.outputShape = Number(params[0].shape()[-1])
    // TODO: self.input_shape = Number(params[0].shape()[0])
  }

}

export class ReLU extends Model {
  static $: IAsyncConstructor = ReLU
  layerType = 'relu'

  static async get(
    id: string
  ) {
    let type = await Model.getModelType(id)

    Model.assertLayerType(type, this)

    return new this(AsyncInstance, id)
  }

  static async create() {
    let id = await Model.createModel(this)

    return new this(AsyncInstance, id)
  }
}

export class Dropout extends Model {
  static $: IAsyncConstructor = Dropout
  layerType = 'dropout'

  static async get(
    id: string
  ) {
    let type = await Model.getModelType(id)

    Model.assertLayerType(type, this)

    return new this(AsyncInstance, id)
  }

  static async create(
    rate = 0.5
  ) {
    let id = await Model.createModel(this, String(rate))

    return new this(AsyncInstance, id)
  }
}

export class Sigmoid extends Model {
  static $: IAsyncConstructor = Sigmoid
  layerType = 'sigmoid'

  static async get(
    id: string
  ) {
    let type = await Model.getModelType(id)

    Model.assertLayerType(type, this)

    return new this(AsyncInstance, id)
  }

  static async create() {
    let id = await Model.createModel(this)

    return new this(AsyncInstance, id)
  }
}

export class Softmax extends Model {
  static $: IAsyncConstructor = Softmax
  layerType = 'softmax'

  static async get(
    id: string
  ) {
    let type = await Model.getModelType(id)

    Model.assertLayerType(type, this)

    return new this(AsyncInstance, id)
  }

  static async create(
    dim = 1
  ) {
    let id = await Model.createModel(this, String(dim))

    return new this(AsyncInstance, id)
  }
}

export class LogSoftmax extends Model {
  static $: IAsyncConstructor = LogSoftmax
  layerType = 'logsoftmax'

  static async get(
    id: string
  ) {
    let type = await Model.getModelType(id)

    Model.assertLayerType(type, this)

    return new this(AsyncInstance, id)
  }

  static async create(
    dim = 1
  ) {
    let id = await Model.createModel(this, String(dim))

    return new this(AsyncInstance, id)
  }
}

export class Log extends Model {
  static $: IAsyncConstructor = Log
  layerType = 'log'

  static async get(
    id: string
  ) {
    let type = await Model.getModelType(id)

    Model.assertLayerType(type, this)

    return new this(AsyncInstance, id)
  }

  static async create() {
    let id = await Model.createModel(this)

    return new this(AsyncInstance, id)
  }
}

export class Tanh extends Model {
  static $: IAsyncConstructor = Tanh
  layerType = 'tanh'

  static async get(
    id: string
  ) {
    let type = await Model.getModelType(id)

    Model.assertLayerType(type, this)

    return new this(AsyncInstance, id)
  }

  static async create() {
    let id = await Model.createModel(this)

    return new this(AsyncInstance, id)
  }
}

export class MSELoss extends Model {
  static $: IAsyncConstructor = MSELoss
  layerType = 'mseloss'

  static async get(
    id: string
  ) {
    let type = await Model.getModelType(id)

    Model.assertLayerType(type, this)

    return new this(AsyncInstance, id)
  }

  static async create() {
    let id = await Model.createModel(this)

    return new this(AsyncInstance, id)
  }

  async forward(
    input: Tensor,
    target: Tensor
  ) {
    let self = this
    self.ready()

    return controller.sendJSON(self.cmd({
      functionCall: 'forward',
      tensorIndexParams: [input.id, target.id]
    }), 'FloatTensor' /*delete_after_use=false*/)
  }
}

export class NLLLoss extends Model {
  static $ : IAsyncConstructor = NLLLoss
  layerType = 'nllloss'

  static async get(
    id: string
  ) {
    let type = await Model.getModelType(id)

    Model.assertLayerType(type, this)

    return new this(AsyncInstance, id)
  }

  static async create() {
    let id = await Model.createModel(this)

    return new this(AsyncInstance, id)
  }


  async forward(
    input: Tensor,
    target: Tensor
  ) {
    let self = this
    self.ready()

    return controller.sendJSON(self.cmd({
      functionCall: 'forward',
      tensorIndexParams: [input.id, target.id]
    }), 'FloatTensor' /*delete_after_use=false*/)
  }
}

export class CrossEntropyLoss extends Model {
  static $ : IAsyncConstructor = CrossEntropyLoss
  layerType: 'crossentropyloss'

  // TODO: backward() to be implemented: grad = target - prediction
  // TODO: backward(): until IntegerTensor is available assume a one-hot vector is passed in.

  static async get(
    id: string
  ) {
    let type = await Model.getModelType(id)

    Model.assertLayerType(type, this)

    return new this(AsyncInstance, id)
  }

  static async create(
    dim = 1
  ) {
    let id = await Model.createModel(this, String(dim))

    return new this(AsyncInstance, id)
  }

  async forward(
    input: Tensor,
    target: Tensor
  ) {
    let self = this
    self.ready()

    return controller.sendJSON(self.cmd({
      functionCall: 'forward',
      tensorIndexParams: [input.id, target.id]
    }), 'FloatTensor' /*delete_after_use=false*/)
  }
}

export class Categorical_CrossEntropy extends Model {
  static $ : IAsyncConstructor = Categorical_CrossEntropy
  layerType: 'categorical_crossentropy'

  static async get(
    id: string
  ) {
    let type = await Model.getModelType(id)

    Model.assertLayerType(type, this)

    return new this(AsyncInstance, id)
  }

  static async create() {
    let id = await Model.createModel(this)

    return new this(AsyncInstance, id)
  }

  async forward(
    input: Tensor,
    target: Tensor
  ) {
    let self = this
    self.ready()

    return controller.sendJSON(self.cmd({
      functionCall: 'forward',
      tensorIndexParams: [input.id, target.id]
    }), 'FloatTensor' /*delete_after_use=false*/)
  }
}
